### PR TITLE
Plugin batch systems

### PR DESCRIPTION
**Pull request by evan-wehi**
This supersedes "Allow dynamic addition of new BatchSystem (#1374)"

This changes allows the registration of new implementations of AbstractBatchSystem. There is a registry for factories to create new BatchSystems (toil.batchSystems.registry) and a seperate registry for adding command line options (toil.batchSystems.options). The registries are pre-populated with the exiting BatchSystem implementations. The reason for separating options and factories is that command line options can be present even if the dependencies for the batch system are not. (Arguably, you might not want command line options present if the can't be used).

The motivating use case is for sites (or even individual users) to be able to add BatchSystems that can implement local policies or to implement BatchSystems that are not generally useful. It also decouples the concrete BatchSystem implementations from common.py.

New BatchSystem can be added programmatically as follows:

```
from toil.batchSystems.registry import addBatchSystemFactory
from toil.batchSystems.options import addOptionsDefinition
    
def drmaaBatchSystemFactory():
    from wehi_pipeline.batchSystems.drmaaBatchSystem import DrmaaBatchSystem
    return DrmaaBatchSystem
    
addBatchSystemFactory('drmaa', drmaaBatchSystemFactory)
       
def addOptions(addOptionFn):
    addOptionFn("--jobQueue", dest="jobQueue", default=None,
            help=("A job queue (used by the DRMAA batch system"))
        
addOptionsDefinition(addOptions)
```

### CODE REVIEW

**Comment by cket** [comment]
Jenkins, test this please

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 1]
```diff
@@ -572,6 +572,12 @@ def executorLost(self, driver, executorId, slaveId, status):
         log.warning("Executor '%s' lost.", executorId)
 
 
+    @classmethod
+    def setOptions(cl, config):
+        super.setOptions(config)
```
Comment (cket) [inline src/toil/batchSystems/mesos/batchSystem.py]: I think this should be `super(MesosBatchSystem, self).setOptions(config)` - that seems to be part of the reason the jenkins build is failing

Comment (evan-wehi) [inline src/toil/batchSystems/mesos/batchSystem.py]: Sorry - lack of familiarity with Python. I've pushed a fix for this. It looks like the PR updates automatically. 

I don't see jenkins output.

**Comment by cket** [comment]
Jenkins, ok to test

**Comment by cket** [comment]
@evan-wehi Thanks again for the code - I’d love to get this feature into Toil! Right now it looks like there are a couple bugs holding Jenkins up. It looks like the default values passed for batch system options aren’t being propagated:
```
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/batchSystems/batchSystemTest.py", line 113, in setUp
    self.batchSystem = self.createBatchSystem()
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/batchSystems/batchSystemTest.py", line 536, in createBatchSystem
    maxDisk=1001)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/batchSystems/parasol.py", line 54, in __init__
    command = config.parasolCommand
AttributeError: 'Config' object has no attribute ‘parasolCommand'
```
There is analogous output for single machine & the `scale` option. The other issue I’m seeing is:
```
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/batchSystems/batchSystemTest.py", line 280, in testJobConcurrency
    Job.Runner.startToil(root, options)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/job.py", line 738, in startToil
    return toil.start(job)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 622, in start
    self._batchSystem = self.createBatchSystem(self.config)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 771, in createBatchSystem
    return batchSystemClass(**kwargs)
TypeError: __init__() takes exactly 6 arguments (5 given)
```
Probably the easiest fix for this is to change Mesos’ constructor to match the other batch systems & just extract the masterAddress attribute from the config.

**Comment by evan-wehi** [comment]
@cket

I’ve fixed the Attribute errors by having the batch system tests add the appropriate attributes to the config. Previously, this was done in common.py but I’ve removed batch system specific code from common.py. I think this is the right solution because it is the responsibility of the batch system to know what its own options are.

I’ve fixed the TypeError in the manner you suggested, although really the factories should know how to construct their own classes.

make test tests=src/toil/test/batchSystems/batchSystemTest.py::SingleMachineBatchSystemTest::testRunJobs

now passes for me.

Thanks for your patience with this.

Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 20 Dec 2016, at 11:05 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@evan-wehi<https://github.com/evan-wehi> Thanks again for the code - I’d love to get this feature into Toil! Right now it looks like there are a couple bugs holding Jenkins up. It looks like the default values passed for batch system options aren’t being propagated:

Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/batchSystems/batchSystemTest.py", line 113, in setUp
    self.batchSystem = self.createBatchSystem()
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/batchSystems/batchSystemTest.py", line 536, in createBatchSystem
    maxDisk=1001)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/batchSystems/parasol.py", line 54, in __init__
    command = config.parasolCommand
AttributeError: 'Config' object has no attribute ‘parasolCommand'


There is analogous output for single machine & the scale option. The other issue I’m seeing is:

Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/batchSystems/batchSystemTest.py", line 280, in testJobConcurrency
    Job.Runner.startToil(root, options)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/job.py", line 738, in startToil
    return toil.start(job)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 622, in start
    self._batchSystem = self.createBatchSystem(self.config)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 771, in createBatchSystem
    return batchSystemClass(**kwargs)
TypeError: __init__() takes exactly 6 arguments (5 given)


Probably the easiest fix for this is to change Mesos’ constructor to match the other batch systems & just extract the masterAddress attribute from the config.

**Comment by cket** [comment]
@evan-wehi, looks like that fixed some of the problems. I’m still seeing the config issue with `AttributeError: 'Config' object has no attribute 'scale’` or `AttributeError: 'Config' object has no attribute 'masterAddress’` in the following tests. 
```
src.toil.test.sort.sortTest.SortTest.testAwsMesos
src.toil.test.sort.sortTest.SortTest.testFileMesos
src.toil.test.batchSystems.batchSystemTest.MesosBatchSystemJobTest.testJobConcurrency
src.toil.test.mesos.mesosTest.MesosTest.test_hello_world
src.toil.test.mesos.mesosTest.MesosTest.test_stress_good
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testDeferralWithConcurrentEncapsulation
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testDeferralWithFailureAndEncapsulation
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testRestart
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testSplitRootPackages
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testUserTypesInJobFunctionArgs
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testUserTypesInJobFunctionArgs
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testConcurrencyDynamic
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testConcurrencyStatic
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testPromiseRequirementRaceStatic
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testPromisesWithJobStoreFileObjects
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testPromisesWithNonCachingFileStore
src.toil.test.batchSystems.batchSystemTest.MaxCoresSingleMachineBatchSystemTest.test
```

**Comment by evan-wehi** [comment]
@cket

I’ve had a another crack at this and Config inserts default configuration for the builtin batch systems (preserving the initial API). It passes the SingleMachineBatchSystemTest but I don’t know how to run your other tests.


Evan.
--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 23 Dec 2016, at 7:55 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@evan-wehi<https://github.com/evan-wehi>, looks like that fixed some of the problems. I’m still seeing the config issue with AttributeError: 'Config' object has no attribute 'scale’ or AttributeError: 'Config' object has no attribute 'masterAddress’ in the following tests.

src.toil.test.sort.sortTest.SortTest.testAwsMesos
src.toil.test.sort.sortTest.SortTest.testFileMesos
src.toil.test.batchSystems.batchSystemTest.MesosBatchSystemJobTest.testJobConcurrency
src.toil.test.mesos.mesosTest.MesosTest.test_hello_world
src.toil.test.mesos.mesosTest.MesosTest.test_stress_good
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testDeferralWithConcurrentEncapsulation
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testDeferralWithFailureAndEncapsulation
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testRestart
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testSplitRootPackages
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testUserTypesInJobFunctionArgs
src.toil.test.src.hotDeploymentTest.HotDeploymentTest.testUserTypesInJobFunctionArgs
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testConcurrencyDynamic
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testConcurrencyStatic
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testPromiseRequirementRaceStatic
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testPromisesWithJobStoreFileObjects
src.toil.test.src.promisedRequirementTest.MesosPromisedRequirementsTest.testPromisesWithNonCachingFileStore
src.toil.test.batchSystems.batchSystemTest.MaxCoresSingleMachineBatchSystemTest.test

**Comment by cket** [comment]
jenkins, test this please

**Comment by benedictpaten** [comment]
@cket I'm seeing what looks like flappy tests on a few of the PRs. Do we have a jenkins problem?

**Comment by cket** [comment]
@benedictpaten the CGCloud integration tests fail a couple times a week whenever pypi is down, but barring those external dependencies the tests have been pretty stable

**Comment by cket** [comment]
@evan-wehi this is passing all the regular tests, but it's hanging on our AWS integration tests. I'll take a deeper look into this and let you if I find anything.

**Comment by evan-wehi** [comment]
Thanks. I'll be back at work in a couple of days.

--
+61 423 000 246

On 10 Jan 2017, at 08:14, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@evan-wehi<https://github.com/evan-wehi> this is passing all the regular tests, but it's hanging on our AWS integration tests. I'll take a deeper look into this and let you if I find anything.

-

**Comment by cket** [comment]
@evan-wehi I haven't found anything that would cause the tests to hang yet.. 

Jenkins, test this please.

**Comment by cket** [comment]
@evan-wehi one of the PRs merged recently introduced conflicts here so you will have to rebase/merge from master. I also recently refactored the integration tests which may have also inadvertently fixed whatever problem was causing them to hang on this branch - when you rebase everything may just magically work.

**Comment by evan-wehi** [comment]
@cket - done, sorry for the delay.

Evan.


--
Evan Thomas
Research Computing Scientist
Computational Biology
The Walter and Eliza Hall Institute of Medical Research
Internal: 2323 | Mobile: +61423000246 | Skype: evanathomas | Email: thomas.e@wehi.edu.au<mailto:thomas.e@wehi.edu.au>




On 26 Jan 2017, at 9:39 am, Christopher Ketchum <notifications@github.com<mailto:notifications@github.com>> wrote:


@evan-wehi<https://github.com/evan-wehi> one of the PRs merged recently introduced conflicts here so you will have to rebase/merge from master. I also recently refactored the integration tests which may have also inadvertently fixed whatever problem was causing them to hang on this branch - when you rebase everything may just magically work.

**Comment by evan-wehi** [comment]
I thought this was merged since I have not heard anything since January! I can't see jenkins output. What should I do to progress this?

**Comment by mr-c** [comment]
@evan-wehi The build seems to have timed out. I don't have permissions to restart the build, sorry.

**Comment by mr-c** [comment]
FYI, there are new conflicts with the master branch. Fixing those would trigger another build.

**Comment by ejacox** [comment]
jenkins, test this please

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by ejacox** [comment]
We are having a problem seeing Jenkins results. You can see the test results here:
http://jenkins.cgcloud.info/view/All/builds (build #2844)

Here is one of the errors:
src/toil/batchSystems/torque.py:36: in <module>
    class TorqueBatchSystem(AbstractGridEngineBatchSystem):
src/toil/batchSystems/torque.py:40: in TorqueBatchSystem
    class Worker(AbstractGridEngineBatchSystem.Worker):
src/toil/batchSystems/torque.py:56: in Worker
    self._version = self._pbsVersion()
E   NameError: name 'self' is not defined


Also, I fixed the conflict in common.py to try to trigger the build. @evan-wehi could you review the changes please. The default configuration was inline rather than a function call.

**Comment by ejacox** [comment]
Ignore the error above. That is for a different PR.

**Comment by ejacox** [comment]
The Jenkins UI is back. The links to the failed tests should work.

**Comment by mr-c** [comment]
@evan-wehi 
```
Traceback (most recent call last):
  File "/home/sort.py", line 189, in <module>
    main()
  File "/home/sort.py", line 186, in main
    memory=sortMemory), options)
  File "/usr/local/lib/python2.7/dist-packages/toil/job.py", line 742, in startToil
    return toil.start(job)
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 653, in start
    self._batchSystem = self.createBatchSystem(self.config)
  File "/usr/local/lib/python2.7/dist-packages/toil/common.py", line 798, in createBatchSystem
    return batchSystemClass(**kwargs)
  File "/usr/local/lib/python2.7/dist-packages/toil/batchSystems/mesos/batchSystem.py", line 93, in __init__
    self.masterAddress = config.masterAddress
AttributeError: 'Config' object has no attribute 'masterAddress'
```

**Comment by evan-wehi** [comment]
Thanks @mr-c. Not sure what has happened here - it looks like I was in mid commit at some point 6 months ago and never finished. I've tested on my local system and it seems good and pushed to this branch. I notice a there seem to be a couple of batch systems in progress. The whole point of this PR was to avoid the need to have built in batch systems. I didn't do it but they could all now be removed from base toil.

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by evan-wehi** [comment]
I not really familiar with this process. Should I be doing something here? I can't see the test output.

**Comment by jessebrennan** [comment]
@evan-wehi It looks the the integration test failure is due to a test hanging. Jenkins then times out and produces no output. This is currently being addressed in #1738. 

In the mean time, the best we have is the [console output](http://jenkins.cgcloud.info/job/toil-pull-request-integration-tests/1190/console). From here it looks to me like all of the integration tests are hanging except the `PremptableDeficitCompensationTest`.

The other [Azure JobStore failure](http://jenkins.cgcloud.info/job/toil-pull-requests/2891/testReport/junit/src.toil.test.src.fileStoreTest/CachingFileStoreTestWithAzureJobStore/testCacheEvictionFailCase/) seems unrelated. I think it is worth testing again to see if that disappears.

**Comment by evan-wehi** [comment]
Ok, seems to have failed again. Can someone give me the error messages?

**Comment by mr-c** [comment]
@evan-wehi 

```
======== 1 passed, 571 deselected, 2 pytest-warnings in 811.61 seconds =========
INFO:__main__:Test keyword UtilsTest and testAWSProvisionerUtils passed successfully
Build timed out (after 240 minutes). Marking the build as failed.
```

Was the only relevant past I saw, but I'm not a Toil logs expert.

**Comment by ejacox** [comment]
I will put this PR on my machine and run it locally. This could be related to PR #1731.

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def __init__(self, config, maxCores, maxMemory, maxDisk, masterAddress):
         self.jobQueues = JobQueue()
 
         # Address of the Mesos master in the form host:port where host can be an IP or a hostname
-        self.masterAddress = masterAddress
+        self.masterAddress = config.masterAddress
```
Comment (jessebrennan) [inline src/toil/batchSystems/mesos/batchSystem.py]: Name of destination in config is inconsistent. Here it is `masterAddress`. [Here](https://github.com/BD2KGenomics/toil/pull/1392/files#diff-b601d50a6eca36eb78f4bc23a332e93cR33) `mesosMasterAddress` is used.

[Here](https://github.com/BD2KGenomics/toil/pull/1392/files#diff-b601d50a6eca36eb78f4bc23a332e93cR84) `masterAddress` is also used.

I'm pretty sure that changing this to 
```
self.masterAddress = config.mesosMasterAddress
```
will solve at least some of the problems.

Comment (evan-wehi) [inline src/toil/batchSystems/mesos/batchSystem.py]: Thanks for looking at it - much appreciated. Python is awesome. I'll push up that change and if it still fails I'll wait for #1738.

**Comment by ejacox** [comment]
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/sort/sortTest.py", line 177, in testAwsMesos
    self._toilSort(jobStoreLocator=self._awsJobStore(), batchSystem="mesos")
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/test/sort/sortTest.py", line 122, in _toilSort
    toil.start(firstJob)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 690, in start
    self._batchSystem = self.createBatchSystem(self.config)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/common.py", line 835, in createBatchSystem
    return batchSystemClass(**kwargs)
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/batchSystems/mesos/batchSystem.py", line 93, in __init__
    self.masterAddress = config.mesosMasterAddress
AttributeError: 'Config' object has no attribute 'mesosMasterAddress'

**Comment by ejacox** [comment]
I was able to reproduce some of the errors locally:
python -m pytest src/toil/test/batchSystems/batchSystemTest.py::MesosBatchSystemTest

**Comment by evan-wehi** [comment]
@ejacox Thanks for taking the trouble to look at this for me. This change was in response to the comment above. I've installed mesos on my laptop but the test above does not reproduce the problem for me. I'm now trying `make test`.

**Comment by evan-wehi** [comment]
I've spent hours on this. I completely rebuilt my environment and went through the documented setup, including the extra steps for mesos and mac OSX. I can't get the mesos tests to run, it claims mesos is not installed.  

@ejacox If it's not tool much trouble can you rerun the tests for me. No hurry, I will be away for a few days.

**Comment by ejacox** [comment]
Thank you for the effort @evan-wehi. I will rerun the test. If I have time, I will try to take a look at the code again today. Otherwise, I should have some time on Monday.

**Comment by ejacox** [comment]
Jenkins, test this please.

**Comment by ejacox** [comment]
Jenkins, test this please.

**Comment by ejacox** [comment]
The AWS tests are failing, but they are failing elsewhere, too. I can't even login to the AWS console. I will rerun the tests when AWS is working again.

**Comment by evan-wehi** [comment]
Thanks again for looking at this for me. If I have time later in the week I’ll setup a better test environment for myself.


Evan.

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by ejacox** [comment]
Jenkins, test this please

**Comment by ejacox** [comment]
Thank you for all of your effort @evan-wehi !

**Comment by evan-wehi** [comment]
@ejacox Thanks for pushing this through and @mr-c and @cket for help along the way!

**Pull request merged**